### PR TITLE
Enhance Lineage Recorder to accept location for a dataset

### PR DESCRIPTION
**Pull request by snehalakshmisha**
Jira: https://cdap.atlassian.net/browse/CDAP-19518

### CODE REVIEW

File: hydrator-common/src/main/java/io/cdap/plugin/common/Asset.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Represents a dataset with a FQN which is a fully-qualified unique identifier for a dataset
+ * and the location of the dataset.
+ */
+public class Asset {
+    private final String fqn;
```
Comment (yaojiefeng) [inline hydrator-common/src/main/java/io/cdap/plugin/common/Asset.java]: nit - indentation should be 2 characters

File: hydrator-common/src/main/java/io/cdap/plugin/common/LineageRecorder.java [hunk 2]
```diff
@@ -36,11 +36,16 @@
 public class LineageRecorder {
 
   private final BatchContext context;
-  private final String dataset;
+  private Asset asset;
```
Comment (yaojiefeng) [inline hydrator-common/src/main/java/io/cdap/plugin/common/LineageRecorder.java]: make this `final`

File: hydrator-common/src/main/java/io/cdap/plugin/common/LineageRecorder.java [hunk 3]
```diff
@@ -36,11 +36,16 @@
...
 
   public LineageRecorder(BatchContext context, String dataset) {
     this.context = context;
-    this.dataset = dataset;
+    this.asset = new Asset(dataset, null);
```
Comment (yaojiefeng) [inline hydrator-common/src/main/java/io/cdap/plugin/common/LineageRecorder.java]: do not pass null to a map, pass a `Collections.emptyMap()` or simply use the single parameter constructor

File: hydrator-common/src/main/java/io/cdap/plugin/common/Asset.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    private final Map<String, String> properties;
+
+    public Asset(String fqn) {
+        this.fqn = fqn;
+        this.properties = new HashMap<>();
```
Comment (yaojiefeng) [inline hydrator-common/src/main/java/io/cdap/plugin/common/Asset.java]: use `Collections.emptyMap()`

File: hydrator-common/src/main/java/io/cdap/plugin/common/Asset.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Represents a dataset with a FQN which is a fully-qualified unique identifier for a dataset
+ * and the location of the dataset.
+ */
+public class Asset {
+    private final String fqn;
+    private final Map<String, String> properties;
```
Comment (yaojiefeng) [inline hydrator-common/src/main/java/io/cdap/plugin/common/Asset.java]: Here I think we can expose the location information as a field, we can just change that to a Map for `Endpoint` in `LineageRecorder`, if you have a Map here and have the plugin to construct this Map, you won't guarantee each plugin will use the correct key for DC to use

**Review by yaojiefeng — COMMENTED** [review]
couple comments

File: hydrator-common/src/main/java/io/cdap/plugin/common/Asset.java (hunk 5 above)
Comment (snehalakshmisha) [reply hydrator-common/src/main/java/io/cdap/plugin/common/Asset.java]: Sounds right

File: hydrator-common/src/main/java/io/cdap/plugin/common/LineageRecorder.java (hunk 3 above)
Comment (snehalakshmisha) [reply hydrator-common/src/main/java/io/cdap/plugin/common/LineageRecorder.java]: Removed the map

File: hydrator-common/src/main/java/io/cdap/plugin/common/Asset.java (hunk 4 above)
Comment (snehalakshmisha) [reply hydrator-common/src/main/java/io/cdap/plugin/common/Asset.java]: Removed the map

File: hydrator-common/src/main/java/io/cdap/plugin/common/Asset.java [hunk 6]
```diff
@@ -15,46 +15,45 @@
...
  * Represents a dataset with a FQN which is a fully-qualified unique identifier for a dataset
  * and the location of the dataset.
  */
 public class Asset {
-    private final String fqn;
-    private final Map<String, String> properties;
+  private final String fqn;
+  private final String location;
+
+  public static final String DEFAULT_LOCATION = "unknown";
```
Comment (yaojiefeng) [inline hydrator-common/src/main/java/io/cdap/plugin/common/Asset.java:24]: nit - move static variable to top of the class

File: hydrator-common/src/main/java/io/cdap/plugin/common/LineageRecorder.java [hunk 7]
```diff
@@ -100,8 +101,13 @@ public void recordRead(String operationName, String operationDescription, List<S
...
-    EndPoint endPoint = EndPoint.of(context.getNamespace(), asset.getFqn(), asset.getProperties());
-    context.record(Collections.singletonList(new FieldWriteOperation(operationName, operationDescription, endPoint,
-                                                                     fields)));
+    context.record(Collections.singletonList(new FieldWriteOperation(operationName, operationDescription,
+                                                                     getEndPoint(), fields)));
+  }
+
+  private EndPoint getEndPoint() {
+    Map<String, String> properties = new HashMap<>();
+    properties.put(Constants.Reference.LOCATION, asset.getLocation());
```
Comment (yaojiefeng) [inline hydrator-common/src/main/java/io/cdap/plugin/common/LineageRecorder.java]: nit - use `Collectons.singletonMap()`

**Review by yaojiefeng — APPROVED** [review]
minor comments, LGTM

**Pull request merged**